core/org.eclipse.cdt.core/parser/org/eclipse/cdt/internal/core/dom/parser/ArithmeticConversion.java

```diff
@@ -201,8 +201,8 @@ private final IType convert(IType type1, IType type2) {
 			return unsignedType;
 		}
 
-		// The signed has the higher rank.
-		if (signedRank.ordinal() > unsignedRank.ordinal()) {
+		// The signed has the higher rank, check if signed type can represent all unsigned values
+		if (fitsIntoType(signedType, unsignedType)) {
```

Comment:
Is there a test we can add to the testsuite to verify this?

Reply:
There is an option to use one of `testTypePromotion_*` tests as starting point to probe the type of binary expression.
I commented on #265 that the test I'd like to have instead did not quite worked because binary expression evaluation did not performed as expected (e.g. 64bit case (1LL - 2UL) has ULL type but still evaluates less than zero). Debugging that confirms the type of binary expression value is now correct so `testTypePromotion_*`-style probe of expression type should succeed.
I also asked on the ticket if there is a real life test case for this which I believe would be preferred in test suite.

Reply:
Added test case, both C and C++ now pass.